src/data/sigma/order.lean

```diff
@@ -181,5 +195,42 @@ instance bounded_order [partial_order ι] [bounded_order ι] [Π i, preorder (α
   bounded_order (Σₗ i, α i) :=
 { .. lex.order_bot, .. lex.order_top }
 
+instance [preorder ι] [densely_ordered ι] [Π i, nonempty (α i)] [Π i, preorder (α i)]
+  [Π i, densely_ordered (α i)] :
+  densely_ordered (Σₗ i, α i) :=
+⟨begin
+  simp_rw sigma.lex.lt_def,
+  rintro ⟨i, a⟩ ⟨j, b⟩ (h | ⟨(rfl : i = j), h⟩),
+  { obtain ⟨k, hi, hj⟩ := exists_between h,
+    obtain ⟨c⟩ : nonempty (α k) := infer_instance,
+    exact ⟨⟨k, c⟩, or.inl hi, or.inl hj⟩ },
+  { obtain ⟨c, ha, hb⟩ := exists_between h,
+    exact ⟨⟨i, c⟩, or.inr ⟨rfl, ha⟩, or.inr ⟨rfl, hb⟩⟩ }
+end⟩
+
+instance densely_ordered_of_no_max_order [preorder ι] [Π i, preorder (α i)]
+  [Π i, densely_ordered (α i)] [Π i, no_max_order (α i)] :
+  densely_ordered (Σₗ i, α i) :=
+⟨begin
+  simp_rw sigma.lex.lt_def,
+  rintro ⟨i, a⟩ ⟨j, b⟩ (h | ⟨(rfl : i = j), h⟩),
+  { obtain ⟨c, ha⟩ := exists_gt a,
+    exact ⟨⟨i, c⟩, or.inr ⟨rfl, ha⟩, or.inl h⟩ },
+  { obtain ⟨c, ha, hb⟩ := exists_between h,
+    exact ⟨⟨i, c⟩, or.inr ⟨rfl, ha⟩, or.inr ⟨rfl, hb⟩⟩ }
+end⟩
+
+instance densely_ordered_of_no_min_order [preorder ι] [Π i, preorder (α i)]
+  [Π i, densely_ordered (α i)] [Π i, no_min_order (α i)] :
+  densely_ordered (Σₗ i, α i) :=
+⟨begin
+  simp_rw sigma.lex.lt_def,
+  rintro ⟨i, a⟩ ⟨j, b⟩ (h | ⟨(rfl : i = j), h⟩),
+  { obtain ⟨c, hb⟩ := exists_lt b,
+    exact ⟨⟨j, c⟩, or.inl h, or.inr ⟨rfl, hb⟩⟩ },
+  { obtain ⟨c, ha, hb⟩ := exists_between h,
+    exact ⟨⟨i, c⟩, or.inr ⟨rfl, ha⟩, or.inr ⟨rfl, hb⟩⟩ }
+end⟩
```

Comment:
I tried to use order_dual and this is what I end up with:
```
lemma order_dual.densely_ordered_iff {α} [l : has_lt α] : densely_ordered α ↔ densely_ordered αᵒᵈ :=
⟨@order_dual.densely_ordered α _,
  by { convert @order_dual.densely_ordered αᵒᵈ _, unfreezingI { cases l }, refl }⟩

lemma lex_swap {r : ι → ι → Prop} {s : Π i, α i → α i → Prop} {a b : Σ i, α i} :
  lex r s a b ↔ lex r.swap (λ i, (s i).swap) b a :=
by split; { rintro (⟨a, b, h⟩ | ⟨a, b, h⟩), exacts [left b a h, right b a h] }

instance densely_ordered_of_no_min_order [preorder ι] [Π i, preorder (α i)]
  [Π i, densely_ordered (α i)] [Π i, no_min_order (α i)] :
  densely_ordered (Σₗ i, α i) :=
order_dual.densely_ordered_iff.2 begin
  convert @lex.densely_ordered_of_no_max_order ιᵒᵈ (λ i, (α i)ᵒᵈ) _ _ _ _,
  apply congr_arg has_lt.mk,
  ext x y, apply lex_swap,
end
```
I'm in favor of keeping your proof, but the first two lemmas seem worth having.